Documentation/release-notes/4509.md

```diff
@@ -1,42 +1,131 @@
+<uses-library android:name="org.apache.http.legacy" android:required="false" />
+```
+
+##### Fallback solution
+
+If updated NuGet package versions are not yet available with the attribute, then
+add the same `UsesLibraryAttribute` directly to the application project instead:
+
+```cs
+[assembly: UsesLibrary("org.apache.http.legacy", false)]
 ```
 
-See the [ProGuard manual][proguard] for details about ProGuard rule syntax.
+##### Background information
+
+The _Missing class: org.apache.http.client_ warning indicates that the app has a
+Java dependency on the Apache HTTP client.  Starting in [Android 6.0 Marshmallow
+(API level 23)][android-6-apache-http], that client was deprecated and moved out
+of _android.jar_ into its own _org.apache.http.legacy.jar_ library.  As a
+result, applications that use Apache HTTP client on Android 6.0 or higher will
+only be able to call the APIs successfully if they declare the use of
+org.apache.http.legacy via a [`<uses-library>`][uses-library] element in
+_AndroidManifest.xml_.  Otherwise, the API calls will fail with exceptions
+similar to:
+
+```txt
+Java.Lang.NoClassDefFoundError: Failed resolution of: Lorg/apache/http/impl/client/DefaultHttpClient
+```
+
+#### How to solve _Missing class: java.lang.ClassValue_
+
+```txt
+R8 : warning : Missing class: java.lang.ClassValue
+R8 : error : Compilation can't be completed because some library classes are missing.
+```
+
+##### Solution
+
+Typically, this warning is safe to suppress:
+
+ 1. Add a new text file to the project, named for example _proguard-rules.pro_.
+
+ 2. Set the contents of the file to:
+
+    ```txt
+    -dontwarn java.lang.ClassValue
+    ```
+
+ 3. Edit the _.csproj_ file in a text editor and add `--pg-conf
+    proguard-rules.pro` to the `AndroidR8ExtraArguments` MSBuild property, for
+    example by adding the following lines:
+
+    ```xml
+    <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+      <AndroidR8ExtraArguments>--pg-conf proguard-rules.pro</AndroidR8ExtraArguments>
+    </PropertyGroup>
+    ```
+
+##### Background information
+
+The _Missing class: java.lang.ClassValue_ warning indicates that the app
+references a library that uses `java.lang.ClassValue`.  Typically this happens
+when the app references a library that was built to run on multiple Java
+platforms.  On Android, the `java.lang.ClassValue` type does not exist in the
+[java.lang package][java-lang-android].
+
+One known scenario that produces this warning is when an app references the
+Xamarin.Firebase.Firestore NuGet package.  That package depends on the
+cross-platform Google Guava library.  Guava uses the `java.lang.ClassValue` type
+only on platforms where it is available, so it is safe to suppress the _Missing
+class: java.lang.ClassValue_ warning in that case.
+
+A future version of the Xamarin.Guava.Google NuGet package will provide the
+`-dontwarn java.lang.ClassValue` R8 rule to suppress the warning automatically.
```

Comment:
I started work on adding the appropriate configuration file to Xamarin.Guava.Google: <https://github.com/xamarin/XamarinComponents/pull/830>.

Along the way, I realized that the Xamarin.Guava.Google NuGet package had recently become a little unusual in its NuGet dependency story due to the fix of using `@(EmbeddedReferenceJar)` to get `guava.jar` from the NuGet package into dependent bindings projects and then into apps. That was going to make it awkward to include the R8 configuration file. So I set up a new fix to make the NuGet dependencies work as expected again: <https://github.com/xamarin/XamarinComponents/pull/831>.

(At the moment, <https://github.com/xamarin/XamarinComponents/pull/830> is not yet rebased on <https://github.com/xamarin/XamarinComponents/pull/831>.)

If we decide that is OK to change `<R8/>` to pass `@(ProguardConfiguration)` items to R8 in multidex-only mode, then <https://github.com/xamarin/XamarinComponents/pull/830> can be simplified even more.